File: lib/csharp/microsoft.bot.builder.solutions/microsoft.bot.builder.solutions/Middleware/SetSpeakMiddleware.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

namespace Microsoft.Bot.Builder.Solutions.Middleware
{
    using System;
    using System.Threading;
    using System.Threading.Tasks;
    using System.Xml;
    using System.Xml.Linq;
    using Microsoft.Bot.Schema;

    /// <summary>
    /// Set Speech Synthesis Markup Language (SSML) on an Activity's Speak property with locale and voice input.
    /// </summary>
    public class SetSpeakMiddleware : IMiddleware
    {
        private const string DefaultLocale = "en-US";

        private const string DefaultVoiceFont = "Microsoft Server Speech Text to Speech Voice (en-US, JessaNeural)";

        private static string _locale;

        private static string _voiceFont;

        private static readonly XNamespace NamespaceURI = @"https://www.w3.org/2001/10/synthesis";

        public SetSpeakMiddleware(string locale = DefaultLocale, string voiceName = DefaultVoiceFont)
        {
            _locale = locale;
            _voiceFont = voiceName;
        }

        /// <summary>
        /// If outgoing Activities are messages and using the Direct Line Speech channel, decorate the Speak property with an SSML formatted string.
        /// </summary>
        /// <param name="context">The Bot Context object.</param>
        /// <param name="next">The next middleware component to run.</param>
        /// <param name="cancellationToken">The cancellation token for the task.</param>
        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
        public Task OnTurnAsync(ITurnContext context, NextDelegate next, CancellationToken cancellationToken = default(CancellationToken))
        {
            context.OnSendActivities(async (ctx, activities, nextSend) =>
            {
                foreach (var activity in activities)
                {
                    switch (activity.Type)
                    {
                        case ActivityTypes.Message:
                            activity.Speak = activity.Speak ?? activity.Text;

                            // TODO: Use Microsoft.Bot.Connector.Channels comparison when "directlinespeech" is available
                            if (activity.ChannelId.Equals("directlinespeech"))
                            {
                                activity.Speak = DecorateSSML(activity);
                            }

                            break;
                    }
                }

                return await nextSend().ConfigureAwait(false);
            });

            return next(cancellationToken);
        }

        /// <summary>
        /// Formats an existing string to be formatted for Speech Synthesis Markup Language with a voice font.
        /// </summary>
        /// <param name="activity">Outgoing bot Activity.</param>
        /// <returns>SSML-formatted string to be used with synthetic speech.</returns>
        private static string DecorateSSML(Activity activity)
        {
            if (string.IsNullOrWhiteSpace(activity.Speak))
            {
                return activity.Speak?.Trim();
            }

            XElement rootElement = null;
            try
            {
                rootElement = XElement.Parse(activity.Speak);
            }
            catch (XmlException)
            {
                // Ignore any exceptions. This is effectively a "TryParse", except that XElement doesn't
                // have a TryParse method.
            }

            if (rootElement == null || rootElement.Name.LocalName != "speak")
            {
                // If the text is not valid XML, or if it's not a <speak> node, treat it as plain text.
                rootElement = new XElement(NamespaceURI + "speak", activity.Speak);
            }

            AddAttributeIfMissing(rootElement, "version", "1.0");
            AddAttributeIfMissing(rootElement, XNamespace.Xml + "lang", _locale);
            AddAttributeIfMissing(rootElement, XNamespace.Xmlns + "mstts", "https://www.w3.org/2001/mstts");

            var sayAsElements = rootElement.Elements("say-as");
            foreach (var element in sayAsElements)
            {
                UpdateAttributeIfPresent(element, "interpret-as", "digits", "number_digit");
            }

            // add voice element if absent
            AddVoiceElementIfMissing(rootElement, _voiceFont);

            return rootElement.ToString(SaveOptions.DisableFormatting);
        }

        /// <summary>
        /// Add a new attribute to an XML element.
        /// </summary>
        /// <param name="element">The XML element to update.</param>
        /// <param name="attributeName">The XML attribute name to add.</param>
        /// <param name="attributeValue">The XML attribute value to add.</param>
        private static void AddAttributeIfMissing(XElement element, XName attributeName, string attributeValue)
        {
            var existingAttribute = element.Attribute(attributeName);
            if (existingAttribute == null)
            {
                element.Add(new XAttribute(attributeName, attributeValue));
            }
        }

        /// <summary>
        /// Add a new attribute with a voice property to the parent XML element.
        /// </summary>
        /// <param name="parentElement">The XML element to update.</param>
        /// <param name="attributeValue">The XML attribute to add.</param>
        private static void AddVoiceElementIfMissing(XElement parentElement, string attributeValue)
        {
            try
            {
                var existingVoiceElement = parentElement.Element("voice") ?? parentElement.Element(NamespaceURI + "voice");

                // If an existing voice element is null (absent), then add it. Otherwise, assume the author has set it correctly.
                if (existingVoiceElement == null)
                {
                    var existingNodes = parentElement.DescendantNodes();

                    XElement voiceElement = new XElement("voice", new XAttribute("name", attributeValue));
                    voiceElement.Add(existingNodes);
                    parentElement.RemoveNodes();
                    parentElement.Add(voiceElement);
                }
                else
                {
                    existingVoiceElement.SetAttributeValue("name", attributeValue);
                }
            }
            catch (Exception ex)
            {
                throw new Exception($"Could not add voice element to speak property {ex.Message}");
            }
        }

        /// <summary>
        /// Update an XML attribute if it already exists.
        /// </summary>
        /// <param name="element">The XML element to update.</param>
        /// <param name="attributeName">The XML attribute name to update.</param>
        /// <param name="currentAttributeValue">The current XML attribute's value.</param>
        /// <param name="newAttributeValue">The new XML attribute's value.</param>
        private static void UpdateAttributeIfPresent(XElement element, XName attributeName, string currentAttributeValue, string newAttributeValue)
        {
            var existingAttribute = element?.Attribute(attributeName);
            if (existingAttribute?.Value == currentAttributeValue)
            {
                existingAttribute?.SetValue(newAttributeValue);
            }
        }
    }
}

File: sdk/typescript/libraries/botbuilder-solutions/src/middleware/setSpeakMiddleware.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { Middleware, SendActivitiesHandler, TurnContext } from 'botbuilder';
import { Activity, ActivityTypes, ResourceResponse } from 'botframework-schema';
import { Element, js2xml, xml2js } from 'xml-js';

const DEFAULT_LOCALE = 'en-US';
const DEFAULT_VOICE_FONT = 'Microsoft Server Speech Text to Speech Voice (en-US, JessaNeural)';

/**
 * Set Speech Synthesis Markup Language (SSML) on an Activity's Speak property with locale and voice input.
 */
export class SetSpeakMiddleware implements Middleware {
    private readonly locale: string;
    private readonly voiceFont: string;
    private readonly namespaceURI: string = 'https://www.w3.org/2001/10/synthesis';

    public constructor(locale: string = DEFAULT_LOCALE, voiceName: string = DEFAULT_VOICE_FONT) {
        this.locale = locale;
        this.voiceFont = voiceName;
    }

    /**
     * If outgoing Activities are messages and using the Direct Line Speech channel,
     * decorate the Speak property with an SSML formatted string.
     * @param context The Bot Context object.
     * @param next The next middleware component to run.
     */
    public async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
        const handler: SendActivitiesHandler = async (
            ctx: TurnContext, activities: Partial<Activity>[], nextSend: () => Promise<ResourceResponse[]>
        ): Promise<ResourceResponse[]> => {
            activities.forEach((activity: Partial<Activity>): void => {
                switch (activity.type) {
                    case ActivityTypes.Message: {
                        if (activity.speak === undefined) {
                            activity.speak = activity.text;
                        }

                        // PENDING: Use Microsoft.Bot.Connector.Channels comparison when "directlinespeech" is available
                        if (activity.channelId === 'directlinespeech') {
                            activity.speak = this.decodeSSML(activity);
                        }

                        break;
                    }
                    default:
                }
            });

            return nextSend();
        };

        context.onSendActivities(handler);

        return next();
    }

    private decodeSSML(activity: Partial<Activity>): string {
        if (activity.speak === undefined || activity.speak.trim() === '') {
            return '';
        }

        let rootElement: Element | undefined = this.elementParse(activity.speak);

        if (rootElement === undefined || this.getLocalName(rootElement) !== 'speak') {
            // If the text is not valid XML, or if it's not a <speak> node, treat it as plain text.
            rootElement = this.createBaseElement(activity.speak);
        }

        this.addAttributeIfMissing(rootElement, 'version', '1.0');
        this.addAttributeIfMissing(rootElement, 'xml:lang', `lang${ this.locale }`);
        this.addAttributeIfMissing(rootElement, 'xmlns:mstts', 'https://www.w3.org/2001/mstts');

        // Fix issue with 'number_digit' interpreter
        if (rootElement.elements !== undefined && rootElement.elements[0].elements !== undefined) {
            const sayAsElements: Element[] = rootElement.elements[0].elements.filter((e: Element): boolean => e.name === 'say-as');
            sayAsElements.forEach((e: Element): void => {
                this.updateAttributeIfPresent(e, 'interpret-as', 'digits', 'number_digit');
            });
        }

        // add voice element if absent
        this.addVoiceElementIfMissing(rootElement, this.voiceFont);

        return js2xml(rootElement, { compact: false });
    }

    private elementParse(value: string): Element | undefined {
        try {
            return xml2js(value, { compact: false }) as Element;
        } catch (error) {
            return undefined;
        }
    }

    private getLocalName(element: Element): string | undefined {
        if (element.elements !== undefined && element.elements.length === 1) {
            return element.elements[0].name;
        }

        return undefined;
    }

    private createBaseElement(value: string): Element {
        // creating simple element
        return {
            elements: [
                {
                    type: 'element',
                    name: 'speak',
                    elements: [
                        {
                            type: 'text',
                            text: value
                        }
                    ],
                    attributes: {
                        xmlns: this.namespaceURI
                    }
                }
            ]};
    }

    private addAttributeIfMissing(element: Element, attName: string, attValue: string): void {
        if (element.elements !== undefined && element.elements[0] !== undefined) {
            if (element.elements[0].attributes === undefined) {
                element.elements[0].attributes = {
                    attName: attValue
                };
            } else {
                if (element.elements[0].attributes[attName] === undefined) {
                    element.elements[0].attributes[attName] = attValue;
                }
            }
        }
    }

    private updateAttributeIfPresent(element: Element, attName: string, attOld: string, attNew: string): void {
        if (element.attributes !== undefined && element.attributes[attName] === attOld) {
            element.attributes[attName] = attNew;
        }
    }

    private addVoiceElementIfMissing(element: Element, attValue: string): void {
        try {
            if (element.elements === undefined || element.elements[0].elements === undefined) {
                throw new Error('rootElement undefined');
            }
            const existingVoiceElement: Element | undefined = element.elements[0].elements.find((e: Element): boolean => e.name === 'voice');

            // If an existing voice element is undefined (absent), then add it. Otherwise, assume the author has set it correctly.
            if (existingVoiceElement === undefined) {
                const oldElements: Element[] = JSON.parse(JSON.stringify(element.elements[0].elements));
                element.elements[0].elements = [
                    {
                        type: 'element',
                        name: 'voice',
                        attributes: {
                            name: attValue
                        },
                        elements: oldElements
                    }
                ];
            } else {
                if (existingVoiceElement.attributes !== undefined) {
                    existingVoiceElement.attributes['name'] = attValue;
                }
            }
        } catch (error) {
            throw new Error(`Could not add voice element to speak property: ${ error.message }`);
        }
    }
}

